linera-core/src/client/validator_manager/manager.rs

```diff
@@ -0,0 +1,1268 @@
+    ///
+    /// Each entry contains:
+    /// - Performance score (f64, normalized 0.0-1.0)
+    /// - EMA success rate (f64, 0.0-1.0)
+    /// - Total requests processed (u64)
+    ///
+    /// Useful for monitoring and debugging node performance.
+    pub async fn get_node_scores(&self) -> BTreeMap<ValidatorPublicKey, (f64, f64, u64)> {
+        let nodes = self.nodes.read().await;
+        let mut result = BTreeMap::new();
+
+        for (key, info) in nodes.iter() {
+            let score = info.calculate_score().await;
+            result.insert(*key, (score, info.ema_success_rate, info.total_requests));
+        }
+
+        result
+    }
+
+    /// Wraps a request operation with performance tracking and capacity management.
+    ///
+    /// This method:
+    /// 1. Acquires a request slot (blocks asynchronously until one is available)
+    /// 2. Executes the provided operation with the selected peer
+    /// 3. Measures response time
+    /// 4. Updates node metrics based on success/failure
+    /// 5. Releases the request slot
+    ///
+    /// # Arguments
+    /// - `peer`: The remote node to execute the operation on
+    /// - `operation`: Async closure that performs the actual request with the selected peer
+    ///
+    /// # Behavior
+    /// If no slot is available, this method will wait asynchronously (without polling)
+    /// until another request completes and releases its slot. The task will be efficiently
+    /// suspended and woken by the async runtime using notification mechanisms.
+    async fn track_request<T, F, Fut>(
+        &self,
+        peer: RemoteNode<Env::ValidatorNode>,
+        operation: F,
+    ) -> Result<T, NodeError>
+    where
+        F: FnOnce(RemoteNode<Env::ValidatorNode>) -> Fut,
+        Fut: Future<Output = Result<T, NodeError>>,
+    {
+        let start_time = Instant::now();
+        let public_key = peer.public_key;
+
+        // Acquire request slot
+        self.try_acquire_slot(&public_key).await;
+
+        // Execute the operation
+        let result = operation(peer).await;
+
+        // Update metrics and release slot
+        let response_time_ms = start_time.elapsed().as_millis() as u64;
+        let is_success = result.is_ok();
+        {
+            let mut nodes = self.nodes.write().await;
+            if let Some(info) = nodes.get_mut(&public_key) {
+                info.update_metrics(is_success, response_time_ms);
+                info.release_request_slot().await;
+                let score = info.calculate_score().await;
+                tracing::trace!(
+                    node = %public_key,
+                    address = %info.node.node.address(),
+                    success = %is_success,
+                    response_time_ms = %response_time_ms,
+                    score = %score,
+                    total_requests = %info.total_requests,
+                    "Request completed"
+                );
+            }
+        }
+
+        // Record prometheus metrics
```

Comment:
```suggestion
        // Record Prometheus metrics
```

Reply:
Done in [Typos and small corrections](https://github.com/linera-io/linera-protocol/pull/4752/commits/22acd1673824df1153592f5cd48d1c3ed359dac4).